### PR TITLE
Add hooks framework

### PR DESCRIPTION
**Pull request by tpcstld**
Add the ability to add hooks (both before and after function execution) onto the `redshift_connection` and `activate_pipeline` functions.

Add the execute command onto `dataduct database`. This will execute the SQL query into the redshift cluster specified by your config file.

Documentation for the hooks api is pending and will be added soon. :)

### CODE REVIEW

**Comment by sb2nov** [comment]
Lets run it by PTB as this is a change away from what we converged upon so I want to keep him in the loop.

File: dataduct/utils/hook.py [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+import imp
+
+
+def default_before_hook(*args, **kwargs):
+    """The default before hook, will act like it's not even there
+    """
+    return args, kwargs
+
+
+def default_after_hook(result):
```
Comment (sb2nov) [inline dataduct/utils/hook.py]: result should default to None I think.

File: dataduct/utils/hook.py [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+    config = Config()
+    # Try to load the hook. If it fails, just fallback to default hooks.
+    try:
+        hook_directory = os.path.expanduser(config.hooks['path'])
+        hook_file = os.path.join(hook_directory, hook_name + '.py')
+
+        custom_hook = imp.load_source('custom_hook', hook_file)
+        before_hook = custom_hook.before_hook
+        after_hook = custom_hook.after_hook
+    except Exception:
```
Comment (sb2nov) [inline dataduct/utils/hook.py]: check if file exists instead of try except.

File: dataduct/utils/hook.py [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+
+def hook(hook_name):
+    """The hook decorator creator
+
+    """
+    from dataduct.config import Config
+    config = Config()
+    # Try to load the hook. If it fails, just fallback to default hooks.
+    try:
+        hook_directory = os.path.expanduser(config.hooks['path'])
```
Comment (sb2nov) [inline dataduct/utils/hook.py]: modify the parse_path function and use that.https://github.com/coursera/dataduct/blob/develop/dataduct/utils/helpers.py

File: dataduct/utils/hook.py [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+
+def default_after_hook(result):
+    """The default after hook, will act like it's not even there
+    """
+    return result
+
+
+def hook(hook_name):
+    """The hook decorator creator
+
```
Comment (sb2nov) [inline dataduct/utils/hook.py]: empty line

File: dataduct/utils/hook.py [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+        hook_file = os.path.join(hook_directory, hook_name + '.py')
+
+        custom_hook = imp.load_source('custom_hook', hook_file)
+        before_hook = custom_hook.before_hook
+        after_hook = custom_hook.after_hook
+    except Exception:
+        before_hook = default_before_hook
+        after_hook = default_after_hook
+
+    def hook_deco(func):
```
Comment (sb2nov) [inline dataduct/utils/hook.py]: rename to hook_decorator

File: dataduct/utils/hook.py [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+    """
+    return result
+
+
+def hook(hook_name):
+    """The hook decorator creator
+
+    """
+    from dataduct.config import Config
+    config = Config()
```
Comment (sb2nov) [inline dataduct/utils/hook.py]: check if hook_path in config first

File: dataduct/utils/hook.py [hunk 7]
```diff
@@ -0,0 +1,50 @@
+"""Hook helpers
```
Comment (sb2nov) [inline dataduct/utils/hook.py]: better docstring

File: dataduct/utils/hook.py (hunk 1 above)
Comment (tpcstld) [inline dataduct/utils/hook.py]: Result will never be unfilled, as
```
result = func(*new_args, **new_kwargs)
new_result = after_hook(result)
```
Is the only place its used.

Better to make sure it's always filled imo.

**Comment by sb2nov** [comment]
LGTM

**Pull request merged**